src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -690,8 +691,36 @@ private void HandleSemanticTokens<T>(string id, T semanticTokensParams, string m
                 tokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
             }
 
-            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+            var controlTokens = isFullDocument ? new LinkedList<ControlToken>() : null;
```

Comment:
Nice to see to this covered. We should not spend time adding and computing control tokens for the range.